vertx-config/src/main/java/io/vertx/config/impl/ConfigRetrieverImpl.java

```diff
@@ -0,0 +1,292 @@
+import io.vertx.config.spi.ConfigStore;
+import io.vertx.config.utils.Processors;
+import io.vertx.core.*;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.config.spi.ConfigProcessor;
+import io.vertx.config.spi.ConfigStoreFactory;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Implementation of the {@link ConfigRetriever}.
+ *
+ * @author <a href="http://escoffier.me">Clement Escoffier</a>
+ */
+public class ConfigRetrieverImpl implements ConfigRetriever {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ConfigRetrieverImpl.class);
+
+  private final Vertx vertx;
+  private final List<ConfigurationProvider> providers;
+  private final long scan;
+  private final List<Handler<ConfigChange>> listeners = new ArrayList<>();
+  private final ConfigStreamImpl streamOfConfiguration = new ConfigStreamImpl();
+
+  private JsonObject current = new JsonObject();
+
+  public ConfigRetrieverImpl(Vertx vertx, ConfigRetrieverOptions options) {
+    this.vertx = vertx;
+
+    ServiceLoader<ConfigStoreFactory> storeImpl =
+        ServiceLoader.load(ConfigStoreFactory.class,
+            ConfigStoreFactory.class.getClassLoader());
+
+    Map<String, ConfigStoreFactory> nameToImplMap = new HashMap<>();
+    storeImpl.iterator().forEachRemaining(factory -> nameToImplMap.put(factory.name(), factory));
+    if (nameToImplMap.isEmpty()) {
+      throw new IllegalStateException("No configuration store implementations found on the classpath");
+    }
+
+    // Iterate over the configured `stores` to configuration the stores
+    providers = new ArrayList<>();
+    for (ConfigStoreOptions option : options.getStores()) {
+      String type = option.getType();
+      if (type == null) {
+        throw new IllegalArgumentException(
+            "the `type` entry is mandatory in a configuration store configuration");
+      }
+
+      ConfigStoreFactory factory = nameToImplMap.get(type);
+      if (factory == null) {
+        throw new IllegalArgumentException("unknown configuration store implementation: " +
+            type + " (known implementations are: " + nameToImplMap.keySet() + ")");
+      }
+
+      JsonObject config = option.getConfig();
+      if (config == null) {
+        config = new JsonObject();
+      }
+      ConfigStore store = factory.create(vertx, config);
+
+      String format = option.getFormat() != null ? option.getFormat() : "json";
+      ConfigProcessor processor = Processors.get(format);
+      if (processor == null) {
+        throw new IllegalArgumentException("unknown configuration format: " + format + " (supported formats are: " +
+            Processors.getSupportedFormats());
+      }
+      providers.add(new ConfigurationProvider(store, processor, option.getConfig()));
+    }
+
+    if (options.getScanPeriod() > 0) {
+      this.scan = vertx.setPeriodic(options.getScanPeriod(), l -> scan());
```

Comment:
this should not be called in the constructor, if the constructor is called from a non Vert.x thread, the timer might execute scan()  before the object is constructed.

the constructor should be encapsulated in a static method that calls the setPeriodic() after the object is initialized.